### PR TITLE
Fixed #31962 -- Made SessionMiddleware raise SessionInterrupted when session destroyed while request is processing.

### PR DESCRIPTION
**Pull request by hramezani**
https://code.djangoproject.com/ticket/31962

### CODE REVIEW

**Comment by hramezani** [comment]
Here is the docs test error:
`WARNING: py:exc reference target not found: django.contrib.sessions.exceptions.SessionIntrupted`

Don't know how to fix it :disappointed:

**Comment by timgraham** [comment]
The error is because of a typo: "Intrupted".

**Comment by hramezani** [comment]
> The error is because of a typo: "Intrupted".

Thanks @timgraham, and sorry for that :man_facepalming:

**Review by felixxm — COMMENTED** [review]
@hramezani Thanks for this patch :+1: I pushed edits to docs. This exception should trigger a 400 response (see [comment](https://github.com/django/django/pull/7431/files#r85411976)), currently it's handled as a 500. We also a need test for this.

**Comment by hramezani** [comment]
Thanks @felixxm for the review and edit!
> This exception should trigger a 400 response

How can I handle this exception as a `HttpResponseBadRequest`?

**Comment by felixxm** [comment]
> How can I handle this exception as a HttpResponseBadRequest?

That's a good question. Maybe we can add a new core exception e.g. `BadRequest` (and `SessionInterrupted` as its subclass) which will be handled in `response_for_exception()`, not sure.

File: django/core/handlers/exception.py [hunk 1]
```diff
@@ -76,6 +76,9 @@ def response_for_exception(request, exc):
             exc_info=sys.exc_info(),
         )
 
+    elif isinstance(exc, BadRequest):
+        response = get_exception_response(request, get_resolver(get_urlconf()), 400, exc)
```
Comment (hramezani) [inline django/core/handlers/exception.py:80]: @felixxm Do we need to use `log_response` here?

**Comment by hramezani** [comment]
> > How can I handle this exception as a HttpResponseBadRequest?
> 
> That's a good question. Maybe we can add a new core exception e.g. `BadRequest` (and `SessionInterrupted` as its subclass) which will be handled in `response_for_exception()`, not sure.

I added `BadRequest` and made `SessionInterrupted` subclass of it.
I am not sure that I did the test in the right way.

File: tests/sessions_tests/tests.py [hunk 2]
```diff
@@ -908,3 +910,10 @@ def test_session_load_does_not_create_record(self):
     @unittest.skip("CookieSession is stored in the client and there is no way to query it.")
     def test_session_save_does_not_resurrect_session_logged_out_in_other_context(self):
         pass
+
+
+@override_settings(DEBUG=True, ROOT_URLCONF='sessions_tests.urls')
+class ResponseDeleteSession(TestCase):
+    def test_response_delete_session(self):
+        response = self.client.get('/response_delete_session/')
+        self.assertContains(response, b'<h1>Bad Request (400)</h1>', status_code=400)
```
Comment (felixxm) [inline tests/sessions_tests/tests.py:919]: I moved this test to `handlers` and `view_tests`.

**Review by felixxm — COMMENTED** [review]
@hramezani Thanks :+1:

File: django/core/exceptions.py [hunk 3]
```diff
@@ -71,6 +71,11 @@ class RequestAborted(Exception):
     pass
 
 
+class BadRequest(Exception):
+    """The request is bad and cannot be processed."""
```
Comment (carltongibson) [inline django/core/exceptions.py:75]:
```suggestion
    """The request is malformed and cannot be processed."""
```

Just the "bad" making me wince a little.

**Review by carltongibson — APPROVED** [review]
This looks good. 

I had a _slight_ worry about the status code… 400 _the request is malformed_… feels like the request is fine but it's unauthorized (the exact same request would succeed otherwise). **HOWEVER** I think 400 is probably the right way to go. 

These kind of timing things are difficult. Makes you glad for `ATOMIC_REQUESTS`. 

Thanks @hramezani an @felixxm

**Comment by felixxm** [comment]
Thanks Carlton!

**Pull request merged**